### PR TITLE
Update cloneElement typings

### PR DESCRIPTION
**Pull request by ddayguerrero**
## What am I trying to achieve?
Resolves https://github.com/preactjs/preact/issues/2363

Updating `cloneElement` type definitions for the `vnode` param and its returned value.

## What approach did I use?
Mostly following Preact's [contribution guidelines](https://github.com/preactjs/preact/blob/master/CONTRIBUTING.md):

- Updated the function's JSDoc block
- Added a test
- Ran `test:ts` and `karma:test:watch` to make sure tests still pass

### CODE REVIEW

File: src/index.d.ts [hunk 1]
```diff
@@ -223,10 +223,10 @@ declare namespace preact {
 		parent: Element | Document | ShadowRoot | DocumentFragment
 	): void;
 	function cloneElement(
```
Comment (JoviDeCroock) [inline src/index.d.ts]: We could allow a generic here as we use [in createElement](https://github.com/preactjs/preact/pull/2388/files#diff-b739f95ace6bbf2f0fd5e8cf1dab97d8R185)

File: src/index.d.ts [hunk 2]
```diff
@@ -17,7 +17,7 @@ declare namespace preact {
 		/**
 		 * ref is not guaranteed by React.ReactElement, for compatiblity reasons
 		 * with popular react libs we define it as optional too
-		*/
+		 */
```
Comment (JoviDeCroock) [inline src/index.d.ts]: You might have some global prettier config

File: test/browser/cloneElement.test.js [hunk 3]
```diff
@@ -8,7 +8,6 @@ describe('cloneElement', () => {
 		const instance = <Comp prop1={1}>hello</Comp>;
 		const clone = cloneElement(instance);
 
-		expect(clone.prototype).to.equal(instance.prototype);
```
Comment (JoviDeCroock) [inline test/browser/cloneElement.test.js]: Why is this deleted, this should be working

File: test/ts/VNode-test.tsx [hunk 4]
```diff
@@ -116,6 +117,19 @@ describe('VNode TS types', () => {
...
 		expect(types).to.deep.equal(['a', '2', 'b', 'c', 'TestComp']);
 	});
+
+	it('component should work with cloneElement', () => {
+		const comp: VNode = (
+			<SimpleComponent>
+				<div>child 1</div>
+			</SimpleComponent>
+		);
+		const clone = cloneElement(comp);
```
Comment (JoviDeCroock) [inline test/ts/VNode-test.tsx]:
```suggestion
		const clone: VNode = cloneElement(comp);
```

I think this is good to type-check since this would throw if cloneElement would return an `Element`

File: src/index.d.ts [hunk 5]
```diff
@@ -222,11 +222,11 @@ declare namespace preact {
 		vnode: ComponentChild,
 		parent: Element | Document | ShadowRoot | DocumentFragment
 	): void;
-	function cloneElement(
-		vnode: JSX.Element,
+	function cloneElement<P>(
```
Comment (JoviDeCroock) [inline src/index.d.ts]: Because of the nature of this file (no ts-compiler) you'll have to specify the non-generic version as well (comparable to createElement https://github.com/preactjs/preact/pull/2388/files#diff-b739f95ace6bbf2f0fd5e8cf1dab97d8R176)

Comment (ddayguerrero) [reply src/index.d.ts]: Thanks for the quick feedback! 😄

File: src/index.d.ts (hunk 2 above)
Comment (ddayguerrero) [reply src/index.d.ts]: I'm not sure what is causing this, I would have thought that no other prettier config ran from my workspace after the pre-commit's `lint-staged`.

That said, it seems to be consistent with the other comment blocks.

Comment (JoviDeCroock) [reply src/index.d.ts]: All right, no worries ;D looks better either way

File: src/index.d.ts (hunk 5 above)
Comment (ddayguerrero) [reply src/index.d.ts]: Shall we consider adding a test for the generic version?

Comment (JoviDeCroock) [reply src/index.d.ts]: Works for me yes! Would be awesome if you did

**Comment by ddayguerrero** [comment]
Thanks! I addressed the remainder of the comments and rebased the commit history.
